frontend/javascripts/oxalis/view/left-border-tabs/histogram_view.js

```diff
@@ -185,6 +194,98 @@ class Histogram extends React.PureComponent<HistogramProps, HistogramState> {
     500,
   );
 
+  getDataForViewport = async (viewport: OrthoView, layerName: string) => {
+    const state = Store.getState();
+
+    const [curX, curY, curZ] = Dimensions.transDim(
+      Dimensions.roundCoordinate(getPosition(state.flycam)),
+      viewport,
+    );
+    const [halfViewportExtentX, halfViewportExtentY] = getHalfViewportExtentsFromState(
+      state,
+      viewport,
+    );
+
+    const min = Dimensions.transDim(
+      V3.sub([curX, curY, curZ], [halfViewportExtentX, halfViewportExtentY, 0]),
+      viewport,
+    );
+    const max = Dimensions.transDim(
+      V3.add([curX, curY, curZ], [halfViewportExtentX, halfViewportExtentY, 1]),
+      viewport,
+    );
+
+    const resolutionIndex = getRequestLogZoomStep(state);
+    const cuboid = await api.data.getDataFor2DBoundingBox(layerName, { min, max }, resolutionIndex);
+    return cuboid;
+  };
+
+  getClippingValues = async (layerName: string, threshold: number = 0.05) => {
+    const { elementClass } = getLayerByName(Store.getState().dataset, layerName);
+    const [TypedArrayClass] = getConstructorForElementClass(elementClass);
+
+    const [cuboidXY, cuboidXZ, cuboidYZ] = await Promise.all([
+      this.getDataForViewport(OrthoViews.PLANE_XY, layerName),
+      this.getDataForViewport(OrthoViews.PLANE_XZ, layerName),
+      this.getDataForViewport(OrthoViews.PLANE_YZ, layerName),
+    ]);
+    const dataForAllViewports = new TypedArrayClass(
+      cuboidXY.length + cuboidXZ.length + cuboidYZ.length,
+    );
+
+    dataForAllViewports.set(cuboidXY);
+    dataForAllViewports.set(cuboidXZ, cuboidXY.length);
+    dataForAllViewports.set(cuboidYZ, cuboidXY.length + cuboidXZ.length);
+
+    const valueCounts = {};
+    let maxCount = 0;
+    for (let i = 0; i < dataForAllViewports.length; i++) {
+      if (dataForAllViewports[i] !== 0 && dataForAllViewports[i] in valueCounts) {
+        valueCounts[dataForAllViewports[i]] += 1;
+        if (maxCount < valueCounts[dataForAllViewports[i]]) {
+          maxCount = valueCounts[dataForAllViewports[i]];
+        }
+      } else {
+        valueCounts[dataForAllViewports[i]] = 1;
+      }
+    }
+
+    let lowClip = -1;
+    let highClip = -1;
+    const threshValue = (threshold * maxCount).toFixed();
```

Comment:
I think, this threshold is a good starting point, but I noticed that it doesn't work too well, when there are large areas with a single intensity. Maybe this is contrived, but the one RGB DS I tested it with, had a large blue sky which dominated the histogram, even when the sky only covered half of the picture. Then, the `maxCount` is very high (in comparison to other other frequencies which are spread over the remaining 254 values) and practically defines the entire clipping range.
I researched a bit and found a very similar approach, which uses the cumulative sum which sounds like it could help to soften this problem scenario very well.

Have a look at this [SO answer](https://stackoverflow.com/a/56909036/896760). The relevant code is this:

```python
    # Calculate cumulative distribution from the histogram
    accumulator = []
    accumulator.append(float(hist[0]))
    for index in range(1, hist_size):
        accumulator.append(accumulator[index -1] + float(hist[index]))

    # Locate points to clip
    maximum = accumulator[-1]
    clip_hist_percent *= (maximum/100.0)
    clip_hist_percent /= 2.0

    # Locate left cut
    minimum_gray = 0
    while accumulator[minimum_gray] < clip_hist_percent:
        minimum_gray += 1

    # Locate right cut
    maximum_gray = hist_size -1
    while accumulator[maximum_gray] >= (maximum - clip_hist_percent):
        maximum_gray -= 1
```

It basically clips a certain percentage of the entire image's intensity starting from left and right in the histogram using the cumulative/accumulated sum. I think, this is closer to what we need. What do you think? Happy to chat about this today :)

Reply:
I implemented this to the best of my knowledge. @philippotto could you please make sure I'm iterating the accumulator and localHist objects correctly. I'm also open for suggestions on how to do it more efficiently :)